src/inverse/SIA2D/gradient.jl

```diff
@@ -157,44 +169,63 @@ function SIA2D_grad_batch!(θ, simulation::FunctionalInversion)
-                function f_adjoint_rev(dλ, λ, p, τ)
+            if !( (typeof(simulation.parameters.UDE.grad.VJP_method) <: DiscreteVJP) | (typeof(simulation.parameters.UDE.grad.VJP_method) <: EnzymeVJP) | (typeof(simulation.parameters.UDE.grad.VJP_method) <: ContinuousVJP) )
+                throw("VJP method $(simulation.parameters.UDE.grad.VJP_method) is not supported yet.")
+            end
+            f_adjoint_rev = let simulation=simulation, H_itp=H_itp, θ=θ
+                function (dλ, λ, p, τ)
                     t = -τ
                     λ_∂f∂H, _ = VJP_λ_∂SIA∂H(simulation.parameters.UDE.grad.VJP_method, λ, H_itp(t), θ, simulation, t)
                     dλ .= λ_∂f∂H
                 end
-            else
-                throw("VJP method $(simulation.parameters.UDE.grad.VJP_method) is not supported yet.")
             end
 
             ### Definition of callback to introduce contribution of loss function to adjoint
             t_ref_inv = .-reverse(t_ref)
-            stop_condition(λ, t, integrator) = Sleipnir.stop_condition_tstops(λ, t, integrator, t_ref_inv)
-            function effect!(integrator)
-                t = - integrator.t
-                ∂ℓ∂H, ∂ℓ∂θ = backward_loss(
-                    loss_function,
-                    H_itp(t),
-                    H_ref_itp(t),
-                    t,
-                    glacier,
-                    θ,
-                    simulation;
-                    normalization = prod(N) * normalization
+            stop_condition_loss(λ, t, integrator) = Sleipnir.stop_condition_tstops(λ, t, integrator, t_ref_inv)
+            effect_loss! = let loss_function=loss_function, H_itp=H_itp, H_ref_itp=H_ref_itp, glacier=glacier, θ=θ, simulation=simulation, normalization=normalization, N=N
+                function (integrator)
+                    t = - integrator.t
+                    ∂ℓ∂H, ∂ℓ∂θ = backward_loss(
+                        loss_function,
+                        H_itp(t),
+                        H_ref_itp(t),
+                        t,
+                        glacier,
+                        θ,
+                        simulation;
+                        normalization = prod(N) * normalization
                     )
-                integrator.u .= integrator.u .+ simulation.parameters.simulation.step .* ∂ℓ∂H
+                    integrator.u .= integrator.u .+ simulation.parameters.simulation.step .* ∂ℓ∂H
+                end
+            end
+            cb_adjoint_loss = DiscreteCallback(stop_condition_loss, effect_loss!)
+            effect_MB! = let simulation=simulation, glacier=glacier, H_itp=H_itp
+                function (integrator)
+                    t = - integrator.t
+                    λ_∂MB∂H = VJP_λ_∂MB∂H(simulation.parameters.UDE.grad.MB_VJP, integrator.u, H_itp(t), simulation, glacier, t)
+                    integrator.u .= λ_∂MB∂H
```

Comment:
Also following my comment before. I think it makes more sense to do `integrator.u .+= ` and just have the contribution of the adjoint. Am I missing something Alban?